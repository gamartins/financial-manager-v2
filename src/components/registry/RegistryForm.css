.RegistryForm .form-control-grow {
    flex-grow: 2;
}