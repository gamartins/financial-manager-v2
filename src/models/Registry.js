export default class Registry {
  static TYPE = {
    PROVENTO: 0,
    DESCONTO: 1,
  }

  static STATUS = {
    NOT_PAID: 0,
    PAID: 1,
    PAID_FINANCIAL_CASH: 2,
    RESERVERD_SAVINGS: 3,
  }

  static getStatusString(status) {
    switch (status) {
    case Registry.STATUS.NOT_PAID:
      return 'NÃ£o pago'
    case Registry.STATUS.PAID:
      return 'Pago'
    case Registry.STATUS.PAID_FINANCIAL_CASH:
      return 'Depositar em poup'
    case Registry.STATUS.RESERVERD_SAVINGS:
      return 'Depositado em poup'
    default:
      return ''
    }
  }
}
