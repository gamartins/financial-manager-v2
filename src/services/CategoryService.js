const key = 'fmv2:category'

const get = () => {
  const dataList = JSON.parse(localStorage.getItem(key))

  if (!dataList) {
    localStorage.setItem(key, JSON.stringify([]))
    return []
  }

  return dataList
}

const save = dataList => {
  localStorage.setItem(key, JSON.stringify(dataList))
}

export default {
  get,
  save,
}
