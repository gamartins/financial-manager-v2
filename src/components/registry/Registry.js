import React, { Component } from 'react'

import addMonths from 'date-fns/addMonths'

import RegistryForm from './RegistryForm'
import RegistryList from './RegistryList'
import DatePagination from '../date-pagination/DatePagination'
import RegistryService from '../../services/RegistryService'

export default class Registry extends Component {

  static TYPE = {
    PROVENTO: 0,
    DESCONTO: 1,
  }

  constructor(props) {
    super(props)

    this.state = {
      date: new Date(),
      dataList: [],
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: ''
      }
    }
  }

  componentDidMount() {
    this.getAll()
  }

  getAll() {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`

    const dataList = RegistryService.get(dateQuery)
    
    this.setState({ dataList })
  }

  getNext() {
    this.setState(
      (state) => ({ date: addMonths(state.date, 1) }),
      () => this.getAll())
  }

  getPrevious() {
    this.setState(
      (state) => ({ date: addMonths(state.date, -1) }),
      () => this.getAll())
  }

  setFormValue(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index === -1) {
      return
    }

    const newFormValues = { ...this.state.dataList[index] }
    
    this.setState({ formValues: newFormValues })
  }

  resetFormValue() {
    this.setState({
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: '',
      }
    })
  }

  save(values) {
    return values.id ? this.edit(values) : this.add(values)
  }

  add(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const dataList = [ ...this.state.dataList ]
    dataList.push({
      id: Math.ceil((Math.random() * 100000)),
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
    })

    RegistryService.save(dataList, dateQuery)
    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  edit(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const index = this.state.dataList.findIndex(item => item.id === values.id)
    const dataList = [ ...this.state.dataList ]
    dataList[index] = {
      id: values.id,
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
    }

    RegistryService.save(dataList, dateQuery)
    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  remove(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index !== -1) {
      this.state.dataList.splice(index, 1)
      this.setState({ dataList: this.state.dataList })
      this.resetFormValue()
    }    
  }

  render() {
    return (
      <div>
        <RegistryForm
          initialValues={this.state.formValues}
          saveItem={values => this.save(values)} />
        <DatePagination
          date={this.state.date}
          format={'MMM'}
          previous={() => this.getPrevious()}
          next={() => this.getNext()}/>
        <RegistryList
          dataList={this.state.dataList}
          removeItem={id => this.remove(id)}
          editItem={item => this.setFormValue(item)} />
      </div>
    )
  }
}
