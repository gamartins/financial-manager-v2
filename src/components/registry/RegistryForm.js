import React from 'react'
import { Formik } from 'formik'

import Registry from './Registry'
import CategoryService from '../../services/CategoryService'

export default class RegistryForm extends React.Component {
  
  categorylist = []

  componentDidMount() {
    this.categorylist = CategoryService.get()
  }

  onSubmit(values, actions) {
    this.props.saveItem(values).then(() => actions.resetForm())
  }

  renderForm(props) {
    return (
      <form onSubmit={props.handleSubmit} className="form-inline">
        <div className='mr-3'>
          <input
            className="form-control"
            type="text"
            name="name"
            placeholder="Nome do item"
            onChange={props.handleChange}
            onBlur={props.handleBlur}
            value={props.values.name} />
        </div>
        
        <div className='mr-3'>
          <select
            className="form-control"
            name="type"
            type="number"
            onChange={props.handleChange}
            onBlur={props.handleBlur}
            value={props.values.type}>
            <option value={Registry.TYPE.PROVENTO}>Provento</option>
            <option value={Registry.TYPE.DESCONTO}>Desconto</option>
          </select>
        </div>

        <div className='mr-3'>
          <select
            className="form-control"
            name="category"
            onChange={props.handleChange}
            onBlur={props.handleBlur}
            value={props.values.category}>
            {this.categorylist.map(category => <option key={category.id} value={category.id}>{category.name}</option>)}
          </select>
        </div>
        
        <div className='mr-3'>
          <div className="input-group">
            <div className="input-group-prepend">
              <div className="input-group-text">R$</div>
            </div>
            <input
              className="form-control"
              type="number"
              name="price"
              placeholder="PreÃ§o"
              onChange={props.handleChange}
              onBlur={props.handleBlur}
              value={props.values.price} />
          </div>
        </div>
        
        <button className="btn btn-primary" type="submit">Salvar</button>
      </form>
    )
  }

  render() {
    return (
      <div className="mb-4">
        <Formik
          initialValues={this.props.initialValues}
          enableReinitialize={true}
          onSubmit={(values, actions) => this.onSubmit(values, actions)}>
            { props => this.renderForm(props) }
        </Formik>
      </div>
    )
  }
}
