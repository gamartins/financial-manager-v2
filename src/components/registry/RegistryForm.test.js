import React from 'react';
import { act } from 'react-dom/test-utils';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import RegistryForm from './RegistryForm';
import Registry from './Registry';

jest.mock('../../services/CategoryService')
import CategoryService from '../../services/CategoryService'

Enzyme.configure({ adapter: new Adapter() });

const initialValues = {
  id: null,
  name: '',
  price: 0,
  type: Registry.TYPE.DESCONTO,
  category: ''
}

CategoryService.get.mockImplementation(() => [
  { id: 1, name: 'Mock category 01' },
  { id: 2, name: 'Mock category 02' },
])

describe('<RegistryForm />', () => {
  test('should be created', () => {
    const wrapper = mount(<RegistryForm initialValues={initialValues} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should get a list of categories', () => {
    const wrapper = shallow(<RegistryForm initialValues={initialValues} />)
    expect(wrapper.instance().categorylist).toHaveLength(2)
  })

  test('should show a list of categories in select element', () => {
    const wrapper = mount(<RegistryForm initialValues={initialValues} />)

    expect(wrapper.find('select[name="category"] option')).toHaveLength(2)
    expect(wrapper.find('select[name="category"] option:first-child').text()).toBe('Mock category 01')
  })

  test('should call saveItem prop on save button click', async () => {
    const saveItem = jest.fn(() => Promise.resolve())
    const component = mount(<RegistryForm
      initialValues={initialValues}
      saveItem={values => saveItem(values)}/>)

    await act(async () => {
      component.find('button').simulate('submit', { preventDefault: () => {} })
    })

    expect(saveItem).toHaveBeenCalledWith(initialValues)
  })

})