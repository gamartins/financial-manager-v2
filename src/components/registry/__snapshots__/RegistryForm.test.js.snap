// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegistryForm /> should be created 1`] = `ReactWrapper {}`;
